File: app-ios/Packages/module-ai/Sources/ModuleAI/Application/Models/AIChatApplicationMapper.swift
//
//  AIChatApplicationMapper.swift
//  ModuleAI
//
//  Created by Codex on 2026/05/08.
//

import Foundation

enum AIChatApplicationMapper {
    private static let chartFallbackMessage = "数据分析还在测试阶段，很快就能上线，敬请期待！"
    
    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()
    
    static func makeMessages(from detailList: [DetailModel]) -> [ConversationMessage] {
        detailList.map(mapToMessage)
    }
    
    static func makeChartPayload(from model: HistoryDetailModel) -> ChartPayload? {
        if let list = model.chartCommonVoList, !list.isEmpty {
            return ChartPayload(
                functionName: model.funcType,
                unit: unit(for: model.funcType),
                series: list.map { item in
                    let title = item.name ?? ""
                    return ChartSeries(
                        xAxis: title,
                        labels: [title],
                        values: [item.value ?? 0]
                    )
                },
                emptyMessage: nil
            )
        }
        
        if let list = model.accountAgeGroupVoList, !list.isEmpty {
            if let first = list.first,
               first.chartType == "2",
               let msg = first.msg {
                return ChartPayload(
                    functionName: model.funcType,
                    unit: unit(for: model.funcType),
                    series: [],
                    emptyMessage: msg
                )
            }
            
            return ChartPayload(
                functionName: model.funcType,
                unit: unit(for: model.funcType),
                series: list.map { item in
                    ChartSeries(
                        xAxis: item.name ?? "",
                        labels: item.labelList ?? [],
                        values: item.valueList ?? []
                    )
                },
                emptyMessage: nil
            )
        }
        
        return nil
    }
}

private extension AIChatApplicationMapper {
    static func mapToMessage(_ model: DetailModel) -> ConversationMessage {
        if model.type == .question {
            return ConversationMessage(
                id: messageId(from: model),
                role: .user,
                contentKind: .text,
                text: model.content ?? "",
                intentType: nil,
                chartPayload: nil,
                feedback: .none,
                historyDetailId: model.id,
                functionName: nil
            )
        }
        
        if let chartDetail = decodeHistoryDetail(from: model) {
            if let payload = makeChartPayload(from: chartDetail) {
                if payload.series.isEmpty == false {
                    return ConversationMessage(
                        id: messageId(from: model),
                        role: .assistant,
                        contentKind: .chart,
                        text: "根据您的查询，以下是分析结果:",
                        intentType: nil,
                        chartPayload: payload,
                        feedback: feedback(from: model.isLike),
                        historyDetailId: model.id,
                        functionName: chartDetail.funcType
                    )
                }
                
                return assistantTextMessage(
                    model,
                    text: payload.emptyMessage ?? Self.chartFallbackMessage
                )
            }
            
            return assistantTextMessage(model, text: Self.chartFallbackMessage)
        }
        
        if let functionModel = decodeFunctionModel(from: model) {
            return assistantTextMessage(model, text: functionModel.msg ?? "")
        }
        
        return assistantTextMessage(
            model,
            text: model.content ?? "新版本上线啦，升级后我会变得更聪明，快来体验吧！"
        )
    }
    
    static func assistantTextMessage(_ model: DetailModel, text: String) -> ConversationMessage {
        ConversationMessage(
            id: messageId(from: model),
            role: .assistant,
            contentKind: .text,
            text: text,
            intentType: nil,
            chartPayload: nil,
            feedback: feedback(from: model.isLike),
            historyDetailId: model.id,
            functionName: nil
        )
    }
    
    static func decodeFunctionModel(from model: DetailModel) -> FunctionModel? {
        guard model.type == .answer,
              model.contentType == .ai,
              let content = model.content,
              let data = content.data(using: .utf8) else {
            return nil
        }
        
        guard let dto = try? decoder.decode(HistoryFunctionContentDTO.self, from: data) else {
            return nil
        }
        
        return dto.toDomainModel()
    }
    
    static func decodeHistoryDetail(from model: DetailModel) -> HistoryDetailModel? {
        guard model.type == .answer,
              model.contentType == .chart,
              let content = model.content,
              let data = content.data(using: .utf8) else {
            return nil
        }
        
        return try? decoder.decode(HistoryDetailModel.self, from: data)
    }
    
    static func feedback(from isLike: String?) -> FeedbackState {
        switch isLike {
        case "1":
            return .liked
        case "0":
            return .disliked
        case nil:
            return .none
        default:
            return .unknown
        }
    }
    
    static func messageId(from model: DetailModel) -> String {
        if let id = model.id {
            return "history-detail-\(id)"
        }
        return UUID().uuidString
    }
    
    static func unit(for functionName: FunctionName?) -> ChartUnit {
        switch functionName {
        case .queryStockGroupByOrg, .queryStockGroupByWarehouse:
            return .ton
        default:
            return .currency
        }
    }
}

private struct HistoryFunctionContentDTO: Decodable {
    let historyId: Int?
    let hasTool: Bool?
    let name: FunctionName?
    let msg: String?
    let arguments: FunctionArguments?
    
    enum CodingKeys: CodingKey {
        case historyId, hasTool, name, msg, arguments
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        historyId = try container.decodeIfPresent(Int.self, forKey: .historyId)
        hasTool = try container.decodeIfPresent(Bool.self, forKey: .hasTool)
        name = try container.decodeIfPresent(FunctionName.self, forKey: .name)
        msg = try container.decodeIfPresent(String.self, forKey: .msg)
        arguments = try Self.decodeArguments(from: container, name: name)
    }
    
    func toDomainModel() -> FunctionModel {
        FunctionModel(
            historyId: historyId,
            hasTool: hasTool,
            name: name,
            msg: msg,
            arguments: arguments
        )
    }
}

private extension HistoryFunctionContentDTO {
    static func decodeArguments(
        from container: KeyedDecodingContainer<CodingKeys>,
        name: FunctionName?
    ) throws -> FunctionArguments? {
        guard let name else { return nil }
        
        switch name {
        case .queryArGroupByOrg, .queryArGroupByCustomer, .queryAccountGroupByAge:
            return .basic(try container.decode(BasicQueryModel.self, forKey: .arguments))
        case .querySalesGroupByOrgAndGoodsType, .querySalesGroupByMonth, .querySalesGroupByCustomer,
             .queryPurchaseGroupByOrg, .queryPurchaseGroupByMonth, .queryPurchaseGroupByCustomer:
            return .timeRange(try container.decode(TimeRangeQueryModel.self, forKey: .arguments))
        case .queryStockGroupByOrg, .queryStockGroupByWarehouse, .queryInventoryGroupByOrg,
             .queryInventoryGroupByWarehouse, .queryProcurementGroupByOrg, .queryProcurementGroupByCustomer:
            return .warehouse(try container.decode(WarehouseQueryModel.self, forKey: .arguments))
        case .queryAccountAgeGroupByOrg, .queryAccountAgeGroupByCustomer:
            return .accountAge(try container.decode(AccountAgeQueryModel.self, forKey: .arguments))
        case .queryPerformanceType:
            return .performanceType(try container.decode(PerformanceTypeQueryModel.self, forKey: .arguments))
        default:
            return nil
        }
    }
}

File: app-android/feature/ai-chat/src/main/java/com/aidatainsight/android/feature/aichat/application/AIChatApplicationMapper.kt
package com.aidatainsight.android.feature.aichat.application

import com.aidatainsight.android.core.model.contract.ChartPayload
import com.aidatainsight.android.core.model.contract.ChartSeries
import com.aidatainsight.android.core.model.contract.ChartUnit
import com.aidatainsight.android.core.model.contract.ConversationContentKind
import com.aidatainsight.android.core.model.contract.ConversationMessage
import com.aidatainsight.android.core.model.contract.ConversationRole
import com.aidatainsight.android.core.model.contract.FeedbackState
import com.aidatainsight.android.core.model.contract.FunctionArguments
import com.aidatainsight.android.core.model.contract.FunctionModel
import com.aidatainsight.android.core.model.contract.FunctionName
import com.aidatainsight.android.core.model.contract.HistoryChartDetail
import com.aidatainsight.android.core.model.contract.HistoryContentType
import com.aidatainsight.android.core.model.contract.HistoryDetail
import com.aidatainsight.android.core.model.contract.HistoryDetailType
import com.aidatainsight.android.core.network.client.AIDataInsightHttpClientFactory
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

object AIChatApplicationMapper {
    private const val CHART_FALLBACK_MESSAGE = "数据分析还在测试阶段，很快就能上线，敬请期待！"
    private const val DEFAULT_ASSISTANT_MESSAGE = "新版本上线啦，升级后我会变得更聪明，快来体验吧！"

    fun makeMessages(detailList: List<HistoryDetail>): List<ConversationMessage> {
        return detailList.map(::mapToMessage)
    }

    fun makeChartPayload(model: HistoryChartDetail): ChartPayload? {
        model.chartCommonVoList?.takeIf { it.isNotEmpty() }?.let { list ->
            return ChartPayload(
                functionName = model.funcType,
                unit = unitFor(model.funcType),
                series = list.map { item ->
                    val title = item.name.orEmpty()
                    ChartSeries(
                        xAxis = title,
                        labels = listOf(title),
                        values = listOf(item.value ?: 0.0),
                    )
                },
                emptyMessage = null,
            )
        }

        model.accountAgeGroupVoList?.takeIf { it.isNotEmpty() }?.let { list ->
            val first = list.first()
            if (first.chartType == "2" && first.msg != null) {
                return ChartPayload(
                    functionName = model.funcType,
                    unit = unitFor(model.funcType),
                    series = emptyList(),
                    emptyMessage = first.msg,
                )
            }

            return ChartPayload(
                functionName = model.funcType,
                unit = unitFor(model.funcType),
                series = list.map { item ->
                    ChartSeries(
                        xAxis = item.name.orEmpty(),
                        labels = item.labelList.orEmpty(),
                        values = item.valueList.orEmpty(),
                    )
                },
                emptyMessage = null,
            )
        }

        return null
    }

    fun makeFunctionModel(data: JsonObject): FunctionModel {
        val name = data["name"]?.jsonPrimitive?.contentOrNull?.let(FunctionName::fromRawValue)
        return FunctionModel(
            historyId = data["historyId"]?.jsonPrimitive?.intOrNull,
            hasTool = data["hasTool"]?.jsonPrimitive?.booleanOrNull,
            name = name,
            msg = data["msg"]?.jsonPrimitive?.contentOrNull,
            arguments = decodeFunctionArguments(name, data["arguments"] as? JsonObject),
        )
    }

    fun decodeFunctionContent(content: String): FunctionModel? {
        val element = runCatching {
            AIDataInsightHttpClientFactory.json.parseToJsonElement(content).jsonObject
        }.getOrNull() ?: return null
        return runCatching { makeFunctionModel(element) }.getOrNull()
    }

    fun decodeChartContent(content: String): HistoryChartDetail? {
        return runCatching {
            AIDataInsightHttpClientFactory.json.decodeFromString(
                HistoryChartDetail.serializer(),
                content,
            )
        }.getOrNull()
    }

    fun chartFallbackMessage(): String = CHART_FALLBACK_MESSAGE

    private fun mapToMessage(model: HistoryDetail): ConversationMessage {
        if (model.type == HistoryDetailType.Question) {
            return ConversationMessage(
                id = messageId(model),
                role = ConversationRole.User,
                contentKind = ConversationContentKind.Text,
                text = model.content.orEmpty(),
                intentType = null,
                chartPayload = null,
                feedback = FeedbackState.None,
                historyDetailId = model.id,
                functionName = null,
            )
        }

        val content = model.content
        if (model.type == HistoryDetailType.Answer &&
            model.contentType == HistoryContentType.Chart &&
            content != null
        ) {
            val chartDetail = decodeChartContent(content)
            if (chartDetail != null) {
                val payload = makeChartPayload(chartDetail)
                if (payload != null && payload.series.isNotEmpty()) {
                    return ConversationMessage(
                        id = messageId(model),
                        role = ConversationRole.Assistant,
                        contentKind = ConversationContentKind.Chart,
                        text = "根据您的查询，以下是分析结果:",
                        intentType = null,
                        chartPayload = payload,
                        feedback = feedbackFrom(model.isLike),
                        historyDetailId = model.id,
                        functionName = chartDetail.funcType,
                    )
                }
                return assistantTextMessage(model, payload?.emptyMessage ?: CHART_FALLBACK_MESSAGE)
            }
            return assistantTextMessage(model, CHART_FALLBACK_MESSAGE)
        }

        if (model.type == HistoryDetailType.Answer &&
            model.contentType == HistoryContentType.Ai &&
            content != null
        ) {
            decodeFunctionContent(content)?.let { functionModel ->
                return assistantTextMessage(model, functionModel.msg.orEmpty())
            }
        }

        return assistantTextMessage(model, content ?: DEFAULT_ASSISTANT_MESSAGE)
    }

    private fun decodeFunctionArguments(
        name: FunctionName?,
        arguments: JsonObject?,
    ): FunctionArguments? {
        if (name == null || arguments == null) return null
        val json = AIDataInsightHttpClientFactory.json
        return runCatching {
            when (name.argumentKind) {
                com.aidatainsight.android.core.model.contract.FunctionArgumentKind.Basic ->
                    FunctionArguments.Basic(json.decodeFromJsonElement(com.aidatainsight.android.core.model.contract.BasicQuery.serializer(), arguments))
                com.aidatainsight.android.core.model.contract.FunctionArgumentKind.TimeRange ->
                    FunctionArguments.TimeRange(json.decodeFromJsonElement(com.aidatainsight.android.core.model.contract.TimeRangeQuery.serializer(), arguments))
                com.aidatainsight.android.core.model.contract.FunctionArgumentKind.Warehouse ->
                    FunctionArguments.Warehouse(json.decodeFromJsonElement(com.aidatainsight.android.core.model.contract.WarehouseQuery.serializer(), arguments))
                com.aidatainsight.android.core.model.contract.FunctionArgumentKind.AccountAge ->
                    FunctionArguments.AccountAge(json.decodeFromJsonElement(com.aidatainsight.android.core.model.contract.AccountAgeQuery.serializer(), arguments))
                com.aidatainsight.android.core.model.contract.FunctionArgumentKind.PerformanceType ->
                    FunctionArguments.PerformanceType(json.decodeFromJsonElement(com.aidatainsight.android.core.model.contract.PerformanceTypeQuery.serializer(), arguments))
            }
        }.getOrNull()
    }

    private fun assistantTextMessage(model: HistoryDetail, text: String): ConversationMessage {
        return ConversationMessage(
            id = messageId(model),
            role = ConversationRole.Assistant,
            contentKind = ConversationContentKind.Text,
            text = text,
            intentType = null,
            chartPayload = null,
            feedback = feedbackFrom(model.isLike),
            historyDetailId = model.id,
            functionName = null,
        )
    }

    private fun feedbackFrom(isLike: String?): FeedbackState {
        return when (isLike) {
            "1" -> FeedbackState.Liked
            "0" -> FeedbackState.Disliked
            null -> FeedbackState.None
            else -> FeedbackState.Unknown
        }
    }

    private fun messageId(model: HistoryDetail): String {
        return model.id?.let { "history-detail-$it" } ?: "history-detail-${model.hashCode()}"
    }

    private fun unitFor(functionName: FunctionName?): ChartUnit {
        return when (functionName) {
            FunctionName.QueryStockGroupByOrg,
            FunctionName.QueryStockGroupByWarehouse,
            -> ChartUnit.Ton
            else -> ChartUnit.Currency
        }
    }
}

